File: 6 – Practice/leet code/001. Two Sum.cpp
// https://leetcode.com/problems/add-two-numbers

class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        ListNode *start = new ListNode(0);
        ListNode *res = start;

        int carry = 0;

        while (l1 || l2) {
            int sum = carry;
            if (l1) {
                sum += l1->val;
                l1 = l1->next;
            }
            if (l2) {
                sum += l2->val;
                l2 = l2->next;
            }

            carry = sum / 10;
            
            start->next = new ListNode(sum % 10);
            start = start->next;
        }
        
        if (carry) {
            start->next = new ListNode(carry);
        }

        return res->next;
    }
};

File: 6 – Practice/leet code/001. Two Sum.java
// https://leetcode.com/problems/two-sum/

class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> lookUp = new HashMap<>();
        
        final int SIZE = nums.length;
        for (int i = 0; i < SIZE; ++i) {
            if (lookUp.containsKey(target - nums[i])) {
                return new int[]{i, lookUp.get(target - nums[i])};
            } else {
                lookUp.put(nums[i], i);
            }
        }
        
        return new int[]{-1, -1};
    }
}
